### PR TITLE
[Codegen] Replace LICM with a version that checks trip count

### PR DESCRIPTION
**Pull request by qedawkins**
The upstream LICM pass does no verification that it is safe to hoist an op out of a loop (i.e. the loop has >= 1 trip count). This replaces all uses of LICM in codegen with a version that does this verification. Other phases of the compiler probably should switch as well.

### CODE REVIEW

**Review by kuhar — COMMENTED** [review]
Should we fix the upstream LICM instead? I think hoisting is still fine if the hoisted ops don't have any side/memory effects -- otherwise it sounds like a bug.

**Comment by qedawkins** [comment]
I filed an issue for this here: https://github.com/iree-org/iree/issues/18759

but I need this change sooner than later for this PR: https://github.com/iree-org/iree/pull/18474

**Review by MaheshRavishankar — COMMENTED** [review]
I can stamp this, but might be also worth fixing the upstream hositing for for loop, and keep this just to avoid hoisting from for_all loops.

**Comment by qedawkins** [comment]
This is the last remaining change blocking #18474 (at least for CI, I'm still going to try to get a model sweep done first) and that PR is the fix to another issue and has been brewing for a while so I'd like to land it sooner rather than later.

File: compiler/src/iree/compiler/Codegen/Common/Passes.td
Comment (nirvedhmeshram) [inline compiler/src/iree/compiler/Codegen/Common/Passes.td]: Maybe add a TODO here linking to https://github.com/iree-org/iree/issues/18759

**Review by nirvedhmeshram — APPROVED** [review]
LGTM, but lets wait for approval from Mahesh or Jakub.

File: compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp]: If we know that the loop always executes, we should change the condition for loop hoisting to only check memory effects. Speculation only matters for the cases when the loop doesn't execute.

File: compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp]: Can we move the logic in this pass to a free function and make this pass just run that? There are passes like OptimizeTensorExtractSlice that run loop hoisting. We can either remove those calls or replace them with this version of LICM.

**Review by Groverkss — COMMENTED** [review]
To avoid problems with the word "safe", can we rename this to IREELoopInvariantCodeMotion?

File: compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp]: Are you sure that's the case? e.g. operations with print side effects could happen once rather than multiple times in the loop.

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp]: I haven't studied the details of the speculatability interface though and could be wrong.

Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp]: > Are you sure that's the case? e.g. operations with print side effects could happen once rather than multiple times in the loop.

printing is a memory sideeffect. speculation is for undefined behavior and infinite loops. Although reading on speculation more, a goto does not have memory effects but is not speculatable and should not be hoisted out. 

I'm probably wrong about my earlier comment and you can ignore it :)

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp]: There's a goto op?!

Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp]: I was reading from here: https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/#modeling

But reading more, actually speculation isn't supposed to handle something like goto/longjmp. Anyways, this was just a passing by comment, I can look more into it and change it later.

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp]: Ok I see, yeah changing it later or maybe adding another case to the interface if we have cases that need it.

**Review by MaheshRavishankar — APPROVED** [review]

**Review by Groverkss — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/test/iree_loop_invariant_code_motion.mlir
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/test/iree_loop_invariant_code_motion.mlir:1]: nit: We don't need splits in test

**Review by kuhar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/test/iree_loop_invariant_code_motion.mlir
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/test/iree_loop_invariant_code_motion.mlir:1]: I prefer the splits because if I add prints or `--mlir-print-ir-after`/`--debug` dumps it will keep those prints within the split regions because it runs the command once per split rather than all at once.

**Pull request merged**
